sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala

```diff
@@ -305,6 +306,63 @@ case class HashPartitioning(expressions: Seq[Expression], numPartitions: Int)
     newChildren: IndexedSeq[Expression]): HashPartitioning = copy(expressions = newChildren)
 }
 
+/**
+ * Represents a partitioning where rows are split across partitions based on transforms defined
+ * by `expressions`. `partitionValuesOpt`, if defined, should contain value of partition key(s) in
+ * ascending order, after evaluated by the transforms in `expressions`, for each input partition.
+ * In addition, its length must be the same as the number of input partitions (and thus is a 1-1
+ * mapping), and each row in `partitionValuesOpt` must be unique.
+ *
+ * For example, if `expressions` is `[years(ts_col)]`, then a valid value of `partitionValuesOpt` is
+ * `[0, 1, 2]`, which represents 3 input partitions with distinct partition values. All rows
+ * in each partition have the same value for column `ts_col` (which is of timestamp type), after
+ * being applied by the `years` transform.
```

Comment:
After reading it, I think it's not really a "hash" partitioning, but more like an "identity" partitioning: rows share the same value w.r.t. the clustering expressions(keys) must be in the same partition, and one partition can only have one value.

Shall we adjust the naming accordingly?

Reply:
like `KeyGroupedPartitioning`?

Reply:
Yes, I also think `HashPartitioning` is not accurate here. What do you think of `DataSourcePartitioning` (different from the `DataSourceHashPartitioning` in catalyst), or `DataStoragePartitioning`? `KeyGroupedPartitioning` sounds fine too.

Reply:
I like `KeyGroupedPartitioning` more because it's not limited to data source v2. Technically the existing `HashPartitioning` can also be a `KeyGroupedPartitioning` using hash expression as the key expression.

Reply:
hmm, aren't we discussing renaming `DataSourceHashPartitioning` to `KeyGroupedPartitioning` here?

Reply:
Yea, these two should match

Reply:
Hmm I'm thinking whether we can still use `DataSourcePartitioning` for the V2 interface, following similar naming conventions such as `DataSourceRDD`, `DataSourceScanExec`, etc. As you mentioned above, `KeyGroupedPartitioning` is not limited to V2 sources but rather more general and can be applied to other things like Hive bucketing in future.

Reply:
The problem is that, v2 `Partitioning` is an interface, which already has 2 implementations in this PR, including UnknownPartitioning. If we want to add more partitioning in the future, the naming must follow its semantic, rather than just a general name `DataSourcePartitioning`.

Reply:
Make sense. Let me switch to `KeyGroupedPartitioning` then.